external/webserver/http_client.c

```diff
@@ -740,3 +833,93 @@ int http_send_response(struct http_client_t *client, int status, const char *bod
 	HTTP_FREE(buf);
 	return HTTP_OK;
 }
+
+int http_send_response_with_status_msg(struct http_client_t *client, int status, const char* status_message, const char* body, struct http_keyvalue_list_t *headers)
+{
+	char *buf;
+	int buflen = 0, ret, sndlen;
+	struct http_keyvalue_t *cur = NULL;
+
+	buf = HTTP_MALLOC(HTTP_CONF_MAX_REQUEST_LENGTH);
+	if (buf == NULL) {
+		HTTP_LOGE("Error: Fail to malloc buffer\n");
+		return HTTP_ERROR;
+	}
+#ifdef CONFIG_NETUTILS_WEBSOCKET
+	if (client->ws_state >= MIN_WS_HEADER_FIELD) {
+		unsigned char accept_key[WEBSOCKET_ACCEPT_KEY_LEN] = {0, };
+		websocket_create_accept_key(accept_key, WEBSOCKET_ACCEPT_KEY_LEN, client->ws_key, WEBSOCKET_CLIENT_KEY_LEN);
+		buflen = snprintf(buf, HTTP_CONF_MAX_REQUEST_LENGTH,
+						  "HTTP/1.1 101 Switching Protocols\r\n"
+						  "Upgrade: websocket\r\n"
+						  "Connection: Upgrade\r\n"
+						  "Sec-WebSocket-Accept: %s\r\n\r\n",
+						  accept_key);
+	} else
+#endif
+	{
+		buflen = snprintf(buf, HTTP_CONF_MAX_REQUEST_LENGTH, "HTTP/1.1 %d %s\r\n",
+						  status, status_message);
+		if (headers) {
+			cur = headers->head->next;
+			while (cur != headers->tail) {
+				buflen += snprintf(buf + buflen, HTTP_CONF_MAX_REQUEST_LENGTH - buflen,
+								   "%s: %s\r\n", cur->key, cur->value);
+				cur = cur->next;
+			}
+		} else {
+			// Add content header
+			if (client->keep_alive == 0) {
+				buflen += snprintf(buf + buflen, HTTP_CONF_MAX_REQUEST_LENGTH - buflen,
+                                                                "Connection: close\r\n");
+			} else {
+				buflen += snprintf(buf + buflen, HTTP_CONF_MAX_REQUEST_LENGTH - buflen,
+                                                                "Connection: Keep-Alive\r\n");
+			}
+
+			// Add content type and content length headers
+			if (body) {
+				buflen += snprintf(buf + buflen,
+                                                           HTTP_CONF_MAX_REQUEST_LENGTH - buflen,
+                                                           "Content-type: text/html\r\n"
+                                                           "Content-Length: %d\r\n",
+                                                           strlen(body));
+			}
+		}
+
+		// Append extra CRLF to mark headers done
+		buflen += snprintf(buf + buflen,
+					HTTP_CONF_MAX_REQUEST_LENGTH - buflen, "\r\n");
+
+		// Include response body
+		if (body) {
+			buflen += snprintf(buf + buflen,
+						HTTP_CONF_MAX_REQUEST_LENGTH - buflen,
+						"%s",
+						body);
+		}
+	}
+
+	sndlen = strlen(buf);
```

Comment:
it seems that but is not initialized after malloc. I'm not sure does snprintf append null character to end of string. If not then you should initialize buf.

Reply:
done